File: 1744-number-of-ways-to-form-a-target-string-given-a-dictionary/1744-number-of-ways-to-form-a-target-string-given-a-dictionary.cpp
class Solution {
public:
    const int MOD = 1e9+7;
    int solve(int i, int j, vector<vector<long long>>& freq, string &target, vector<vector<int>>& dp) {
        if (i == target.size())
            return 1;
        if (j == freq[0].size())
            return 0;
        if (dp[i][j] != -1)
            return dp[i][j];
        int not_taken = solve(i, j+1, freq, target, dp)%MOD;
        int taken = (freq[target[i]-'a'][j]*solve(i+1, j+1, freq, target, dp))%MOD;
        return dp[i][j] = (not_taken + taken)%MOD;
    }
    
    int numWays(vector<string>& words, string target) {
        vector<vector<long long>> freq(26, vector<long long>(words[0].size()));
        for(int col=0;col<words[0].size();col++)
        {
            for(auto& word:words)
                freq[word[col]-'a'][col]++;
        }
        vector<vector<int>> dp(target.length()+1, vector<int>(words[0].size()+1, -1));
        return solve(0, 0, freq, target, dp);
    }
};

File: 1744-number-of-ways-to-form-a-target-string-given-a-dictionary/1744-number-of-ways-to-form-a-target-string-given-a-dictionary.java
class Solution {
    int MOD = (int) 1e9 + 7;
    private int recursion(int[][] freq, String target, int[][] memo, int i, int tlen) {
        if (tlen == target.length()) {
            return 1;
        }

        if (i >= freq.length) {
            return 0;
        }

        if (memo[i][tlen] != -1) {
            return memo[i][tlen];
        }

        int take = 0, notTake = 0;

        char tchar = target.charAt(tlen);

        if (freq[i][tchar - 'a'] > 0) {
            take = (int) ((long) freq[i][tchar - 'a'] * recursion(freq, target, memo, i + 1, tlen + 1) % MOD);
        }

        notTake = recursion(freq, target, memo, i + 1, tlen) % MOD;

        memo[i][tlen] = (take + notTake) % MOD;
        return memo[i][tlen];
    }

    public int numWays(String[] words, String target) {
        int m = words[0].length();
        int[][] freq = new int[m][26];

        for (String word : words) {
            for (int i = 0; i < m; i++) {
                freq[i][word.charAt(i) - 'a']++;
            }
        }

        int[][] memo = new int[m][target.length()];
        for (int[] dp : memo) {
            Arrays.fill(dp, -1);
        }

        return recursion(freq, target, memo, 0, 0);
    }
}